packages/core/src/Synchronizer.ts

```diff
@@ -396,211 +543,161 @@ export class Synchronizer extends EventEmitter {
+        if (attesterId !== this.attesterId.toString()) return
         db.create('UserSignUp', {
-            commitment: idCommitment.toString(),
+            commitment,
             epoch,
             attesterId,
-            airdrop,
         })
         return true
     }
 
-    async attestationEvent(event: ethers.Event, db: TransactionDB) {
-        const _epoch = Number(event.topics[1])
-        const _epochKey = BigInt(event.topics[2])
-        const _attester = event.topics[3]
-        const decodedData = this.unirepContract.interface.decodeEventLog(
-            'AttestationSubmitted',
-            event.data
-        )
+    async handleAttestationSubmitted({
+        decodedData,
+        event,
+        db,
+    }: EventHandlerArgs) {
+        const epoch = Number(decodedData.epoch)
+        const epochKey = BigInt(decodedData.epochKey).toString()
+        const attesterId = BigInt(decodedData.attesterId).toString()
+        const posRep = Number(decodedData.posRep)
+        const negRep = Number(decodedData.negRep)
+        if (attesterId !== this.attesterId.toString()) return
 
         const index = `${event.blockNumber
             .toString()
             .padStart(15, '0')}${event.transactionIndex
             .toString()
             .padStart(8, '0')}${event.logIndex.toString().padStart(8, '0')}`
 
-        await this._checkCurrentEpoch(_epoch)
-        await this._checkEpochKeyRange(_epochKey.toString())
+        const currentEpoch = await this.readCurrentEpoch()
+        if (epoch !== Number(currentEpoch.number)) {
+            throw new Error(
+                `Synchronizer: Epoch (${epoch}) must be the same as the current synced epoch ${currentEpoch.number}`
+            )
+        }
+        await this._checkEpochKeyRange(epochKey)
 
-        const attestation = new Attestation(
-            BigInt(decodedData.attestation.attesterId),
-            BigInt(decodedData.attestation.posRep),
-            BigInt(decodedData.attestation.negRep),
-            BigInt(decodedData.attestation.graffiti),
-            BigInt(decodedData.attestation.signUp)
-        )
         db.create('Attestation', {
-            epoch: _epoch,
-            epochKey: _epochKey.toString(),
-            index: index,
-            attester: _attester,
-            attesterId: Number(decodedData.attestation.attesterId),
-            posRep: Number(decodedData.attestation.posRep),
-            negRep: Number(decodedData.attestation.negRep),
-            graffiti: decodedData.attestation.graffiti.toString(),
-            signUp: Number(decodedData.attestation?.signUp),
-            hash: attestation.hash().toString(),
+            epoch,
+            epochKey,
+            index,
+            attesterId,
+            posRep,
+            negRep,
+            graffiti: decodedData.graffiti.toString(),
+            timestamp: decodedData.timestamp.toString(),
+            hash: hash2([posRep, negRep]).toString(),
```

Comment:
do we need this?
or it should be `hash4([posRep, negRep, graffiti, timestamp])`

Reply:
Related #200